### ISSUE TITLE
Identify correct instance when using launching multiple instances

### ISSUE THREAD

**Issue by Duroxxigar** [godotengine/godot-proposals#3357]
### Describe the project you are working on

A networked game in pre-alpha 4.0

### Describe the problem or limitation you are having in your project

When using the new feature to launch multiple instances (pic 1), it is very easy to lose track of which instance belongs to which debug session (pic 2).

![debug](https://user-images.githubusercontent.com/16636969/134939302-cc2df117-31a5-4d5a-9208-ca34633cbc9d.png)

![instances](https://user-images.githubusercontent.com/16636969/134939411-c2339d37-d85e-402e-ab77-edf312701090.png)
 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

In the instance window title bar, display the debug session number, something like, `Session: 2`. This will help prevent this situation altogether.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Read above

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Will be used all the time for games that are testing with multiple instances.

### Is there a reason why this should be core and not an add-on in the asset library?

It's a core editor feature now.

**Comment by chrisb123** [comment]
This will be handy property to expose, currently all the instances are created and stacked on top of each other with the exact same name, there is no way to determine or label which is which session and I have to manually arrange the windows every time they are started

**Comment by Calinou** [comment]
The technical implementation of this feature likely involves adding a `--window-title-prefix` command line argument that can be set to override the project's window title, including any run-time changes it may perform.

**Comment by chrisb123** [comment]
I worked out how to do this in gdscript with a network game I was working on. I'll need to find it

**Comment by opatut** [comment]
> The technical implementation of this feature likely involves adding a --window-title-prefix command line argument that can be set to add a prefix to the project's window title.

I would appreciate it if this would double as a way to identify the different sessions from script, e.g. `--session-index` -> `Engine.get_session_index()` or something along those lines, so I can automatically make one session a server and the other the client on startup in my debug entrypoint scene, such that I end up in a working game without interaction, for even faster testing.

**Comment by Eoin-ONeill-Yokai** [comment]
@opatut @Calinou I recently had to make my own modification to the engine to account for this but would like to know how users feel before attempting an upstream.

I've basically created an instance identifier variable (for instance, `--instance-count 1~4`) which will be added to the command line arguments only when there are multiple instances available. 

So by opatut's recommendation, I could make an `Engine.get_session_index()` option, but would need to look into where the `is_debugger` variables are instantiated to get an idea of how to make this more accessible. Right now, I'm actually use it for my initialization of server / client for debugging. For my server client check, I'm simply looking at all the input args for `--instance-count` and then grab the next variable. If it's 1, I run the server. If it's 2 or more, I run the client instance. I'm fine with either.

With that in mind, users could configure whatever behavior they need from this (including modifying window title states, or overlaying detailed information unique to each instance as an overlay in game) while also having **other** uses for it such as testing server client interactions.

Basically, I have the code ready already to mostly upstream, but would like to clean it up if the main team is interested. Should I create a WIP merge request?

**Comment by chrisb123** [comment]
exposing the session ID with something like Engine.get_session_index() would be ideal, no need for more

**Comment by kdada** [comment]
Multiple instanses may connect to debug server with arbitry order.

I have another implementation for this proposal: Add misc capture group to RemoteDebugger and add misc:set-session-id to set the session id.

**Comment by Mimerme** [comment]
I think adding a feature to seperate the two console instances as well would be helpful as its currently difficult to differentiate between the two while print debugging

**Comment by kuyarruyyy** [comment]
I was scrolling through the docs for a solution and I came across `DisplayServer.window_get_attached_instance_id()` [here](https://docs.godotengine.org/en/stable/classes/class_displayserver.html#class-displayserver-method-window-get-attached-instance-id) and it looked promising. So, I put it in a label, run 2 instances of the project, and got different values between the instances.

Is this what we are looking for? If not, can someone enlighten me on what it actually does and how is it different from what we are trying to achieve?

**Comment by quimnuss** [comment]
Meanwhile, I found OS.get_process_id() which _at least_ lets me distinguish between windows. As a quick and fallible workaround this missing info I'm just doing modulo 2 to choose player 0 or 1 which works most of the time.

There's probably a way to leverage this more robustly.

**Comment by KonstantinosPetrakis** [comment]
One way to handle this is the following (by storing an incremental value to a file): 


```gdbscript
func get_incremental_number():
	var save_path = "user://incremental.tres"
	
	var store_value = func(val): 
		var file = FileAccess.open(save_path, FileAccess.WRITE)
		file.store_64(val)
		file.flush()
		file.close()
	
	var read_value = func():
		if FileAccess.file_exists(save_path):
			await get_tree().create_timer(.5).timeout # Wait for val to be saved
			var file = FileAccess.open(save_path, FileAccess.READ)
			var value = file.get_64()
			file.close()
			return value
		return 0 
	
	var value = await read_value.call()
	store_value.call(value + 1)
	return value


func _ready():
	if await get_incremental_number() % 2 == 0:
		_on_lobby_host_game(Globals.SERVER_PORT)
	else:
		_on_lobby_connect_game("localhost", Globals.SERVER_PORT)
```

It's not the brightest way, but it's used for debugging purposes anyway.

**Comment by ctrlraul** [comment]
Here's a way to give an index/id to the instances running:
![image](https://github.com/godotengine/godot-proposals/assets/47065615/dc33923c-e44b-4bd3-8628-382d855a0a94)

![image](https://github.com/godotengine/godot-proposals/assets/47065615/c28b3be0-0594-4499-939f-62c4db5894f6)

And the code to obtain the value - I'm using C# but the GDScript equivalent shouldn't be too dissimilar:
```csharp
public readonly int InstanceIndex = GetInstanceIndex();

private static int GetInstanceIndex()
{
	const string pattern = @"instance-index=(\d+)";
	
	Regex regex = new Regex(pattern);

	foreach (string arg in OS.GetCmdlineArgs())
	{
		Match match = regex.Match(arg);
		
		if (match.Success)
			return int.Parse(match.Groups[1].Value);
	}

	return -1;
}
```
I settled for a regex patten match, but you can always use a command line arguments parsing library for it.

I hope that's useful to anyone else. 🙂

**Comment by D4r3NPo** [comment]
@JodliDev It's a new feature of the 4.3 : 
![Capture d’écran 2024-06-24 à 17 49 49](https://github.com/godotengine/godot-proposals/assets/46395128/8f7cebc6-6631-40e3-96a4-c420e997d412)

**Comment by JodliDev** [comment]
Oh. Well, I should have figured that out myself. Thank you very much! :)

**Comment by recon472** [comment]
is it possible to set these new settings in 4.3 through gdscript? I'd like to add buttons to my editor which would change the settings in this customize run instances window

**Comment by aaronfranke** [comment]
As of proposal #522 and PR https://github.com/godotengine/godot/pull/65753, you can manually edit the launch arguments to add an instance number if desired. However, it still doesn't show in the title of the window.
